### ISSUE TITLE
Command sometime is rejected

### ISSUE THREAD

**Issue by piotrva** [piotrva/esphome_gree_ac#2]
The behavior was observed, that sometimes (randomly) command is not accepted by the unit.
Sending it second time provides a workaround.

Need to investigate this and add more robust algorithm for applying settings (maybe internal check if settings are applied with a limited number of retires) .

**Comment by gekkehenkie11** [comment]
Ok yeah I see this in my logs too:
[code][20:38:11][D][climate:040]:   Target Temperature: 23.00
[20:38:11][V][sinclair_ac.serial:085]: Requested target teperature change
[20:38:11][V][sinclair_ac:347]: RX: 7E.7E.31.31.00.00.40.00.90.80.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.43.00.45.00.00.7E (52)
[20:38:11][V][sinclair_ac:345]: TX: 7E.7E.2F.01.00.00.00.AF.90.70.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.25 (50)
[20:38:11][V][sinclair_ac:345]: TX: 7E.7E.2F.01.00.00.00.00.90.70.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.76 (50)[/code]

So after sending the command there's just no response from the AC. The data itself doesn't seem to be the problem, exactly the same packet did receive a response shortly after that in my logs. Probably would need to log more of the original device to see if this happens too. But maybe it's also a timing thing, so maybe the original device actually waits a bit longer for a RX? Or maybe it doesnt but just resends the command (as you do in a 'workaround' like you mentioned) ?

**Comment by gekkehenkie11** [comment]
My current theory is that we need to mimic that cmd 0x3 packet. Not sure what triggers it but there are 4 subtypes of this packet, specified by the byte after the instruction byte. Each triggers a different packet type from the AC:

subtype 07: triggers cmd 0x33 from AC
subtype 06: triggers cmd 0x34 from AC
subtype 04: triggers cmd 0x35 from AC
subtype 00: triggers cmd 0x31 from AC

So it keeps repeating these commands 0x3 in the above order.

**Comment by gekkehenkie11** [comment]
Done some more analysis and I'm really starting to think we'll need to implement sending of CMD 0x3. I believe that the airco uart might be doing some reset because it's not receiving those commands and then misses some of the commands.

Actually the 0x3 command is pretty easy:

7e 7e 1a 03 07[1] 00[2] 00[3] 00[4] 00[5] 00[6] 00[7] 00[8] 00[9] 00[10] 00[11] 7e[12] 96[13] 1d[14] 0c[15] 0b[16] 21[17] 00[18] 00[19] 00[20] 00[21] 00[22] 01[23] 00[24] 8e[25] 

Here those [13], [14], [15] is the date, it's a deriviation of year, month, day. And then follows the time (hour, minute, sec). And that byte [1] I already outlined above.

Now I just need to find time to implement this, will probably be a while ....

**Comment by gekkehenkie11** [comment]
Actually found a moment to reverse engineer some more and found that at least one timer that gets set to 10 seconds and then does something with the 0x3 packet. I currently believe it sets the 'sub type' status (5th byte) back to 0x7 and initiates the whole chain of sending. I saw that the changes of that byte (7 -> 6 ->  4 ->0) happen in the handlers of the response of those packets. So currently it seems that this chain starts every 10 seconds, subtype is set to 7, and then the other 3 subtypes get set by the 3 handlers of the response packets from the AC (0x33, 0x34, 0x35).

Anyway now REALLY done for a while, really need to focus on other stuff now ...

**Comment by gekkehenkie11** [comment]
Ok this indeed definitely is a different issue. To recreate, just keep changing the temperature like every second. You will hear the AC beep each time. However sometimes when you for example change from 24 to 23, you see in HA the value 23 flash for a few milliseconds and then you will see it back to what it was before, the 24. And also the module then just resends that 24 value instead of the 23 value. So it will send the exact same command as the one before, but the AC will beep ofcourse because 0xAF was set...

So this seems to be something close to HA. Honestly I'm not an expert there, haven't done much HA stuff yet. Will see if I have time to look into this, if not, maybe you can have a look @piotrva ?

**Comment by gekkehenkie11** [comment]
So for example this is the log, I changed temp, saw it flash back in HA immediately and there's not even a trace of the new temp in the logs, so it was never sent downstream:

```
[09:26:05][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.AF.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.06 (50)
[09:26:05][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:05][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.57 (50)
[09:26:05][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:05][D][climate:396]: 'Keuken' - Sending state:
[09:26:05][D][climate:399]:   Mode: COOL
[09:26:05][D][climate:407]:   Custom Fan Mode: 1 - Low
[09:26:05][D][climate:416]:   Swing Mode: OFF
[09:26:05][D][climate:419]:   Current Temperature: 25.00°C
[09:26:05][D][climate:425]:   Target Temperature: 21.00°C
[09:26:05][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.08.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.5F (50)
[09:26:05][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:05][D][climate:396]: 'Keuken' - Sending state:
[09:26:05][D][climate:399]:   Mode: COOL
[09:26:05][D][climate:407]:   Custom Fan Mode: 1 - Low
[09:26:05][D][climate:416]:   Swing Mode: OFF
[09:26:05][D][climate:419]:   Current Temperature: 25.00°C
[09:26:05][D][climate:425]:   Target Temperature: 21.00°C
[09:26:05][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.08.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.5F (50)
[09:26:06][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:06][D][climate:396]: 'Keuken' - Sending state:
[09:26:06][D][climate:399]:   Mode: COOL
[09:26:06][D][climate:407]:   Custom Fan Mode: 1 - Low
[09:26:06][D][climate:416]:   Swing Mode: OFF
[09:26:06][D][climate:419]:   Current Temperature: 25.00°C
[09:26:06][D][climate:425]:   Target Temperature: 21.00°C
[09:26:06][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.08.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.5F (50)
[09:26:06][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:06][D][climate:396]: 'Keuken' - Sending state:
[09:26:06][D][climate:399]:   Mode: COOL
[09:26:06][D][climate:407]:   Custom Fan Mode: 1 - Low
[09:26:06][D][climate:416]:   Swing Mode: OFF
[09:26:06][D][climate:419]:   Current Temperature: 25.00°C
[09:26:06][D][climate:425]:   Target Temperature: 21.00°C
[09:26:06][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.08.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.5F (50)
[09:26:06][D][climate:011]: 'Keuken' - Setting
[09:26:06][D][climate:040]:   Target Temperature: 21.00
[09:26:06][V][sinclair_ac.serial:086]: Requested target teperature change
[09:26:06][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:06][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.AF.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.06 (50)
[09:26:07][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:07][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.57 (50)
[09:26:07][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:07][D][climate:396]: 'Keuken' - Sending state:
[09:26:07][D][climate:399]:   Mode: COOL
[09:26:07][D][climate:407]:   Custom Fan Mode: 1 - Low
[09:26:07][D][climate:416]:   Swing Mode: OFF
[09:26:07][D][climate:419]:   Current Temperature: 25.00°C
[09:26:07][D][climate:425]:   Target Temperature: 21.00°C
[09:26:07][V][sinclair_ac:366]: TX: 7E.7E.2F.01.00.00.00.00.91.50.02.02.40.00.00.08.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.02.00.00.00.00.00.5F (50)
[09:26:07][V][sinclair_ac:368]: RX: 7E.7E.31.31.00.00.40.00.91.50.02.02.40.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.1D.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.00.42.00.46.00.00.6C (52)
[09:26:07][D][climate:396]: 'Keuken' - Sending state:

```

**Comment by gekkehenkie11** [comment]
Yeah that's definitely the problem. if you delete the "publish_state"
[this->publish_state();](https://github.com/piotrva/esphome_gree_ac/blob/60444cb0bf13fc56c57d9b29a04694dc77912458/components/sinclair_ac/esppac_cnt.cpp#L592)

Then the problem is gone. However of course the values wont update in HA if you update them with the remote control. Not sure yet about the best way to fix that ...

**Comment by gekkehenkie11** [comment]
Ok I also fixed that in my fork. What I did is I save the packet when the user sends a new setting via HA. Then before sending the publish_state back to HA, I check if the packet is the same as what the user changed it to. If so there's no need to send it back to HA. So we only send packets to HA that are not initiated via HA.

Everything works fine now!

**Comment by gekkehenkie11** [comment]
Correct.

**Comment by gekkehenkie11** [comment]
Actually since I was coding online in the webbrowser there are like a zillion updates on my fork LOL. But if you filecompare the esppac_cnt.cpp then the relevant action is in:

void SinclairACCNT::handle_packet()

So I compare the packet that it wants to push back to home assistant (via  'publish_state') to what we've gotten from home assistant. And if it's a packet we got from home assistant in the first place, there's no need to publish it back! So that's the idea. I compare bytes 4-6 and 8-11 but would actually verify if more needs to be compared. I already saw that we should not compare byte 7 cause that would give problems (since it sometimes differs between RX and TX).

Also some changes that were initiated by HA, we do need to push back. Like a mode change from COOL to HEAT for example, so that's why I also added the 'reqmodechange' boolean to incorporate that.

**Comment by piotrva** [comment]
In the PR I do not see any changes to reqmodechange

**Comment by gekkehenkie11** [comment]
No, I only have submitted a PR for the first issue. Sorry, really low on time right now

**Comment by piotrva** [comment]
Unfortunatelly after implementing your PR component stopped working at all in my setup...
Need to revert the change

**Comment by gekkehenkie11** [comment]
Ah 1 possibility is that your AC model does not send out 0x31 packets  without receiving 0x1 packets! So in that case the code would cause both the AC and the module to wait for each other! Try to remove the wait for packet condition and only have it check the 300 ms timer. That's how the original Gree module works too! 

So instead of my PR, use:

if (millis() - this->last_packet_sent_ < protocol::TIME_REFRESH_PERIOD_MS)

I guess if your AC works differently than mine, this indeed could be a problem.

Either way, since the original Gree module works exactly like that, I guess we should be doing that too ...  Haven't verified if this could cause any problems with other parts of your code, only 1 way to find out LOL, but I bet this works right away.

**Comment by gekkehenkie11** [comment]
So I'd say, try the above first and see how it performs! If it DOES cause any unexpected behaviour I guess we could just simply add to my PR that if if it's been waiting more than 600 ms, it shoud fire the 0x1 packet anyway ...
